// zipkin/test-types/recorder.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./tsr-declarations");
var lib_1 = require("ts-runtime/lib");
var zipkin_1 = require("zipkin");
var chai_1 = require("chai");
describe('Recorder types', function () {
    describe('BatchRecorder', function () {
        it('should return correct type', function () {
            var batchRecorder = lib_1.default.nullable(lib_1.default.ref("zipkin.BatchRecorder.979108923")).assert(new zipkin_1.BatchRecorder({
                logger: new /** @class */ (function () {
                    function class_1() {
                    }
                    class_1.prototype.logSpan = function (span) {
                    };
                    return class_1;
                }()),
                timeout: 1000,
            }));
            chai_1.expect(batchRecorder.record).to.be.a('function');
        });
    });
    describe('ConsoleRecorder', function () {
        it('should ConsoleRecorder return correct type', function () {
            var consoleRecorder = lib_1.default.nullable(lib_1.default.ref("zipkin.ConsoleRecorder.979108923")).assert(new zipkin_1.ConsoleRecorder());
            chai_1.expect(consoleRecorder.record).to.be.a('function');
        });
    });
});
